File: Project Academy/12. Pagination-test/java/app/src/test/java/com/dicoding/academies/ui/detail/DetailCourseViewModelTest.java
package com.dicoding.academies.ui.detail;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import com.dicoding.academies.data.AcademyRepository;
import com.dicoding.academies.data.source.local.entity.CourseEntity;
import com.dicoding.academies.data.source.local.entity.CourseWithModule;
import com.dicoding.academies.data.source.local.entity.ModuleEntity;
import com.dicoding.academies.utils.DataDummy;
import com.dicoding.academies.vo.Resource;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class DetailCourseViewModelTest {
    private DetailCourseViewModel viewModel;
    private CourseEntity dummyCourse = DataDummy.generateDummyCourses().get(0);
    private String courseId = dummyCourse.getCourseId();
    private ArrayList<ModuleEntity> dummyModules = DataDummy.generateDummyModules(courseId);

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    @Mock
    private AcademyRepository academyRepository;

    @Mock
    private Observer<Resource<CourseWithModule>> observer;

    @Before
    public void setUp() {
        viewModel = new DetailCourseViewModel(academyRepository);
        viewModel.setCourseId(courseId);
    }

    @Test
    public void getCourseWithModule() {
        Resource<CourseWithModule> dummyCourseWithModule = Resource.success(DataDummy.generateDummyCourseWithModules(dummyCourse, true));
        MutableLiveData<Resource<CourseWithModule>> course = new MutableLiveData<>();
        course.setValue(dummyCourseWithModule);

        when(academyRepository.getCourseWithModules(courseId)).thenReturn(course);

        viewModel.courseModule.observeForever(observer);

        verify(observer).onChanged(dummyCourseWithModule);
    }
}

File: Project Academy/12. Pagination-test/kotlin/app/src/test/java/com/dicoding/academies/ui/detail/DetailCourseViewModelTest.kt
package com.dicoding.academies.ui.detail

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.dicoding.academies.data.AcademyRepository
import com.dicoding.academies.data.source.local.entity.CourseWithModule
import com.dicoding.academies.utils.DataDummy
import com.dicoding.academies.vo.Resource
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class DetailCourseViewModelTest {
    private lateinit var viewModel: DetailCourseViewModel
    private val dummyCourse = DataDummy.generateDummyCourses()[0]
    private val courseId = dummyCourse.courseId

    @get:Rule
    var instantTaskExecutorRule = InstantTaskExecutorRule()

    @Mock
    private lateinit var academyRepository: AcademyRepository

    @Mock
    private lateinit var observer: Observer<Resource<CourseWithModule>>

    @Before
    fun setUp() {
        viewModel = DetailCourseViewModel(academyRepository)
        viewModel.setSelectedCourse(courseId)
    }

    @Test
    fun `setSelectedCourse should be success`() {
        val expected = MutableLiveData<Resource<CourseWithModule>>()
        expected.value = Resource.success(DataDummy.generateDummyCourseWithModules(dummyCourse, true))

        `when`(academyRepository.getCourseWithModules(courseId)).thenReturn(expected)

        viewModel.courseModule.observeForever(observer)

        verify(observer).onChanged(expected.value)

        val expectedValue = expected.value
        val actualValue = viewModel.courseModule.value

        assertEquals(expectedValue, actualValue)
    }

    @Test
    fun `setBookmark should not add favorite when courseModule is empty`() {
        viewModel.setBookmark()
        verify(academyRepository, times(0)).setCourseBookmark(dummyCourse, true)
    }

    @Test
    fun `setBookmark should add favorite when courseModule is not empty and not bookmarked`() {
        val expected = MutableLiveData<Resource<CourseWithModule>>()
        expected.value = Resource.success(DataDummy.generateDummyCourseWithModules(dummyCourse, false))

        `when`(academyRepository.getCourseWithModules(courseId)).thenReturn(expected)
        viewModel.courseModule.observeForever(observer)

        doNothing().`when`(academyRepository).setCourseBookmark(dummyCourse, true)

        viewModel.setBookmark()
        verify(academyRepository, times(1)).setCourseBookmark(dummyCourse, true)
    }

    @Test
    fun `setBookmark should remove from favorite when courseModule is not empty and already bookmarked`() {
        val expected = MutableLiveData<Resource<CourseWithModule>>()
        expected.value = Resource.success(DataDummy.generateDummyCourseWithModules(dummyCourse, true))

        `when`(academyRepository.getCourseWithModules(courseId)).thenReturn(expected)
        viewModel.courseModule.observeForever(observer)

        doNothing().`when`(academyRepository).setCourseBookmark(dummyCourse, false)

        viewModel.setBookmark()
        verify(academyRepository, times(1)).setCourseBookmark(dummyCourse, false)
    }

}